### ISSUE TITLE
HIP error: invalid device function

### ISSUE THREAD

**Issue by AbdullahSK03** [ROCm/ROCm#3401]
I am trying to train a neural network model using my gpu but it is showing errors related to rocm and HIP.

OS: {
NAME="Ubuntu"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
} 

CPU: AMD Ryzen 5 3400G with Radeon Vega Graphics
GPU: AMD Radeon RX 6600 (amdgcn-amd-amdhsa--gfx1032
![Screenshot from 2024-07-06 20-56-37](https://github.com/ROCm/ROCm/assets/121804317/b7508264-3786-458b-9d31-b033a7ec3cad)
)

**Comment by evshiron** [comment]
If you have integrated graphics, make sure you are using the discrete graphics instead of the integrated one.

You can use `rocminfo` or `rocm_agent_enumerator` to list all compatible ROCm devices, and use environment variables like `HIP_VISIBLE_DEVICES=1` to make only the second GPU (excluding CPUs) are visible by the applications, which can be a list with comma as the separator.

Environment variables like `HSA_OVERRIDE_GFX_VERSION=10.3.0` can also be used if RDNA 2 GPUs don't work out of the box.

**Comment by harkgill-amd** [comment]
Hi @AbdullahSK03, were you able to give the suggestion above a try?

**Comment by harkgill-amd** [comment]
@AbdullahSK03, we were able to reproduce this issue internally. As @evshiron mentioned, this is due to HIP using your integrated graphics rather than your discrete graphics. By setting `HIP_VISIBLE_DEVICES=1`, the issue is resolved as only the GPU will be visible to HIP applications. You can find more information about GPU isolation techniques [here](https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html). 

If you feel that this solution does not address your issue completely, feel free to re-open the ticket. Thanks!

**Comment by evshiron** [comment]
@AbdullahSK03 

Have you tried this tutorial?

* https://github.com/alfinauzikri/ROCm-RX6600XT

I don't have a RX 6600 XT, but as tutorials exist it should work if everything is set up properly.

**Comment by AbdullahSK03** [comment]
I have RX6600 non xt variant.

**Comment by harkgill-amd** [comment]
Could you please provide the output of `rocminfo` and `rocm_agent_enumerator` after disabling your integrated graphics?

**Comment by AbdullahSK03** [comment]
Thank you all for the Help. After some trials and errors, I got it to work as per my requirements.

What I did:
Instead of going with the latest I went for older version.
Version: 6.0.0.60000-91~22.04

Leavening details as this might come in handy to someone like me in future.

https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.0.2/tutorial/quick-start.html

My System Hardware Info:
OS: Ubuntu 22.04.4 Jammy Jellyfish
CPU: AMD Ryzen 5 3400G with Vega 11 graphics
RAM: 16 GB 3200MHz (Enabled XMP)
MB: Gigabyte B450M D3-SH
GPU: Asus Radeon RX6600 (non-XT)

more OS detaills:
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.4 LTS
Release:	22.04
Codename:	jammy

Kernel Details:

<code>
(base) abdullah@AbdullahUbuntu:~$ dpkg -l | grep rocm
ii  rocm-core                                  6.0.0.60000-91~22.04                              amd64        Radeon Open Compute (ROCm) Runtime software stack
ii  rocm-hip-libraries                         6.0.0.60000-91~22.04                              amd64        Radeon Open Compute (ROCm) Runtime software stack
ii  rocm-hip-runtime                           6.0.0.60000-91~22.04                              amd64        Radeon Open Compute (ROCm) Runtime software stack
ii  rocm-language-runtime                      6.0.0.60000-91~22.04                              amd64        Radeon Open Compute (ROCm) Runtime software stack
ii  rocm-llvm                                  17.0.0.23483.60000-91~22.04                       amd64        ROCm compiler
ii  rocm-smi-lib                               6.0.0.60000-91~22.04                              amd64        AMD System Management libraries
ii  rocminfo                                   1.0.0.60000-91~22.04                              amd64        Radeon Open Compute (ROCm) Runtime rocminfo tool
</code>

<code>
Linux AbdullahUbuntu 6.5.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Jul 15 16:40:02 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

</code>
